### PR TITLE
Huge Louvain Preparation: more Huge Arrays

### PR DESCRIPTION
**Pull request by knutwalker**
In addition to the existing `HugeLongArray` as replacement for `long[]`:
- `int[] -> HugeIntArray`
- `double[] -> HugeDoubleArray`
- `T[] -> HugeObjectArray<T>`

### CODE REVIEW

**Comment by jexp** [comment]
Can you comment/note somewhere on your perf-test of the Neo4j Batch-Importer Arrays.

File: core/src/main/java/org/neo4j/graphalgo/core/utils/paged/HugeIntArray.java [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+        }
+
+        @Override
+        public long size() {
+            return size;
+        }
+
+        @Override
+        public long release() {
+            if (page != null) {
```
[the remark this answers is not in the document]
Comment (knutwalker) [reply core/src/main/java/org/neo4j/graphalgo/core/utils/paged/HugeIntArray.java]: I think I wanted to not drag the Tracker around in the arrays, but I agree that it is inconsistent with what other release methods do. I'll change it so that the tracker is a parameter of the release method

File: core/src/main/java/org/neo4j/graphalgo/core/utils/paged/HugeLongArray.java [hunk 2]
```diff
@@ -109,67 +117,67 @@
...
+    @Override
+    public void boxedSet(final long index, final Long value) {
+        set(index, value);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void boxedSetAll(final LongFunction<Long> gen) {
```
[the remark this answers is not in the document]
Comment (knutwalker) [reply core/src/main/java/org/neo4j/graphalgo/core/utils/paged/HugeLongArray.java]: boxedMethods are for the test mainly, I'll make them package-private instead of public

**Comment by knutwalker** [comment]
@jexp I addressed most of your notes, I just missed the static methods from the Cursor/init methods.

Regarding benchmarks, I ran this PR on my laptop and got the following results:

```
 mvn -DskipTests -PBenchmark -pl benchmark package
java -jar benchmark/target/benchmark.jar -jvmArgs "-server -Xmx4g -Xms4g -XX:+UseG1GC" -psize=10000000 -psparseness=0.0 -pdistribution=uniform -gc true -rf json -rff long-arrays.json org.neo4j.graphalgo.utils.LongArrayBenchmark
```

[long-arrays.json.zip](https://github.com/neo4j-contrib/neo4j-graph-algorithms/files/3034381/long-arrays.json.zip)

long-arrays

the tested arrays are:

primitve: plain Java `long[]`
huge_paged: `org.neo4j.graphalgo.core.utils.paged.HugeLongArray` with the paging configuration (nested `long[][]`)
huge_paged_cursor: using the HugeCursor instead of get/set
huge_single: `org.neo4j.graphalgo.core.utils.paged.HugeLongArray` with a single page configuration (wrapper around `long[]`)
huge_single_cursor: using the HugeCursor instead of get/set
sparse: `org.neo4j.graphalgo.core.utils.paged.SparseLongArray`, similar to huge_paged, but pages can be null and r/w access includes appropriate checks
offHeap: `org.neo4j.unsafe.impl.batchimport.cache.OffHeapLongArray`
chunked: `org.neo4j.unsafe.impl.batchimport.cache.DynamicLongArray`

some possible observations:
- the JIT can inline the huge-single-page version to be the same as plain primitve access
- the same goes for cursor based access on the huge-paged version, otherwise its 30%-80% slower
- reads for the off-heap array is good, but writes are rather slow and very inconsistent
- sparse array just slightly slower than the huge-paged without cursor (as expected)

results might be something different on other machines with different CPUs, of course.

File: core/src/main/java/org/neo4j/graphalgo/core/utils/paged/HugeArrays.java [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 final class HugeArrays {
 
-    private static final int SINGLE_PAGE_SHIFT = 30;
+    private static final int SINGLE_PAGE_SHIFT = 28;
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/paged/HugeArrays.java]: why this change? was this a bug?

Comment (knutwalker) [reply core/src/main/java/org/neo4j/graphalgo/core/utils/paged/HugeArrays.java]: allocating a large array might trigger a gc cycle, possibly full, to make room for the array (long, doubles, objects require 8 consecutive gigs with a shift of 30) and we have no way of figuring out if this is going to happen – It might also fail even when there seems to be enough heap available and it takes a while to allocate an array that long. With 28, we just bail out earlier to the paged version to not trigger those edge cases that often.
